### DISCUSSION TITLE
Sharing a PGP message that is already encrypted

### DISCUSSION THREAD

**Discussion by bruncanepa** [openpgpjs/openpgpjs#1705] [Q&A]
Hey!

I'm trying to share an encrypted PGP message with multiple users. 

To do this I need to:
1. Get the decrypted `session key` packets from the PGP message using `decryptSessionKeys`.
2. Encrypt the `session key` using `encrypt` and passing as encryption key the `receiver PGP public key`.
3. Create a new message with the `encrypted data packets` and the `encrypted key packets` from step 2.

The problem I'm having, is that I don't know how to get only the `encrypted data packets` from the PGP message to avoid having to encrypt again the data to be shared.

Is this possible? Or I have to create a new `passphrase` to encrypt the PGP message, and then encrypt the `passphrase` with the `receiver PGP public key` ?

Thank you!

**Comment by larabr** [comment]
There is `openpgp.encryptSessionKey()` you can use, passing the decrypted session key data, as well as the new encryption keys.
If you get a binary output, then you can prepend the session key packets to the original encrypted message, and the resulting data can be shared with everyone (it will also include the original encrypted session key).
The following code illustrates what I mean (but I haven't tested it, so I might require some minor changes):

```js
const [sessionKey] = await decryptSessionKeys({ message, decryptionKeys: <originalRecipientKeys> });
const reEncryptedSessionKey = await encryptSessionKey({
  ...sessionKey,
  encryptionKeys: <newRecipientPublicKeys>,
  format: 'binary'
});

const finalBinaryMessage = new Uint8Array([...reEncryptedSessionKey, ...message.write()]);
const finalArmoredMessage = await openpgp.readMessage({ binaryMessage: finalBinaryMessage }).then(parsed => parsed.armor());
```

If you want to filter out the original encrypted session keys, you can use e.g.`message.packets.filterByTag(openpgp.enums.packet.symEncryptedIntegrityProtectedData)` and call `write()` over the returned packet (there cannot be more).
If you filter the packets, be mindful that there are other types of messages being standardized, so filtering only `symEncryptedIntegrityProtectedData` might not work in the future 🙂 

PS: if the original message is signed, and includes an `intended recipient` subpacket in the signature, note that the new people you are sharing the message with _won't_ be able to verify the message (as they are not the intended recipients 😁 ).

**Reply by bruncanepa** [reply]
I understand, thank you for the response.

So is this the best way to share a PGP message that is already encrypted? Encrypting the data with a random key, that is then shared?

**Encrypt**:
1. Create a random key of 32 bytes => `passphrase`.
2. Encrypt with PGP or directly with AES-GCM the `message` using as password the `passphrase` => `encrypted message`
3. Encrypt  the `passphrase` using all users PGP public keys and sign each with my PGP private key => `encrypted passphrase`

**Decrypt** (run by a receiver user)
1. Decrypt `encrypted passphrase` with my PGP public key => `passphrase`
2. Decrypt `encrypted message` with `passphrase` => `message`

**Reply by larabr** [reply]
It depends what the use-case is. The simplest way is to re-encrypt the session key as described before. Note that when encrypting an OpenPGP message with a password, you still have session keys being generated internally, so the system is not more efficient.
If you are using AES-GCM directly in step 2, then I assume the passphrase will be the AES key, in which case, for sharing, you can encrypt it using `encryptSessionKey`, instead of `encrypt`. 

What's the reason to sign the passphrase?

**Reply by bruncanepa** [reply]
The reason to sign the passphrase is to confirm who the sender is. Because if I don't sign the message (that's only encrypted), at least I need to sign the passphrase.

About the use-case: I'm building a [library ](https://github.com/bruncanepa/open-e2ee) to simplify E2EE, so I can use it as a base to build new application on-top of the library.
The features I need are:
- Encrypt / decrypt
- Share encrypted data while verifying the sender

A specific example would be a notes app built on-top of this library:
- I have my notes encrypted/decrypted.
- I want to share an encrypted note with another user using PGP public key.

Maybe the best solution for my library should be:
- Encrypt data using AES-GCM, and encrypt (and  maybe sign ) the AES key with my PGP keys.
- For sharing, I decrypt the AES key, and encrypt with receiver PGP public key.

**Reply by larabr** [reply]
> The reason to sign the passphrase is to confirm who the sender is.

Careful tho, because any of the recipients of message A, can re-use the passphrase and its signature when sending a different message B.
If you want authenticity of the message, you need to sign the message (or a digest). You could generate a digest of the original content, then encrypt+sign that, for each recipient. But the easiest thing would be to just encrypt+sign the original message (OpenPGP.js currently does not support adding the "intended recipient" signature subpacket, so the result will be verifiable by all recipients. Once we add support for the future, it'll be possible to disable it).

**Reply by bruncanepa** [reply]
Oh yes, you are right! Thank you for everything!
